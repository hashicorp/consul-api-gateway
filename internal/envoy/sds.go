package envoy

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"errors"
	"fmt"
	"net"
	"net/url"
	"sync"
	"time"

	secretservice "github.com/envoyproxy/go-control-plane/envoy/service/secret/v3"
	cache "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
	"github.com/envoyproxy/go-control-plane/pkg/log"
	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
	server "github.com/envoyproxy/go-control-plane/pkg/server/v3"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/grpclog"

	"github.com/hashicorp/go-hclog"

	grpcint "github.com/hashicorp/consul-api-gateway/internal/grpc"
	"github.com/hashicorp/consul-api-gateway/internal/metrics"
	"github.com/hashicorp/consul-api-gateway/internal/store"
)

//go:generate mockgen -source ./sds.go -destination ./mocks/sds.go -package mocks CertificateFetcher

const (
	defaultGRPCBindAddress = ":9090"
	defaultShutdownTimeout = 10 * time.Second

	cachedMetricsTimeout = 10 * time.Second
)

var logOnce sync.Once

// CertificateFetcher is used to fetch the CA and server certificate
// that the server should use for TLS
type CertificateFetcher interface {
	SPIFFE() *url.URL
	RootPool() *x509.CertPool
	TLSCertificate() *tls.Certificate
}

// SDSServer wraps a gRPC-based SDS Delta server
type SDSServer struct {
	logger      hclog.Logger
	fetcher     CertificateFetcher
	server      *grpc.Server
	client      SecretClient
	bindAddress string
	protocol    string
	store       store.Store
}

// NEWSDSServer initializes an SDSServer instance
func NewSDSServer(logger hclog.Logger, fetcher CertificateFetcher, client SecretClient, store store.Store) *SDSServer {
	return &SDSServer{
		logger:      logger,
		fetcher:     fetcher,
		client:      client,
		bindAddress: defaultGRPCBindAddress,
		protocol:    "tcp",
		store:       store,
	}
}

// Run starts the SDS server
func (s *SDSServer) Run(ctx context.Context) error {
	childCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	logOnce.Do(func() {
		grpclog.SetLoggerV2(grpcint.NewHCLogLogger(s.logger))
	})

	opts := []grpc.ServerOption{
		grpc.MaxConcurrentStreams(2048),
		grpc.Creds(credentials.NewTLS(&tls.Config{
			GetConfigForClient: func(hello *tls.ClientHelloInfo) (*tls.Config, error) {
				certificate := s.fetcher.TLSCertificate()
				if certificate == nil {
					return nil, errors.New("unable to get server certificate")
				}
				pool := s.fetcher.RootPool()
				if pool == nil {
					return nil, errors.New("unable to get CA pool")
				}
				return &tls.Config{
					Certificates: []tls.Certificate{*certificate},
					ClientCAs:    pool,
					ClientAuth:   tls.RequireAndVerifyClientCert,
				}, nil
			},
			ClientAuth: tls.RequireAndVerifyClientCert,
		})),
		grpc.StreamInterceptor(SPIFFEStreamMiddleware(s.logger, s.fetcher, s.store)),
	}
	s.server = grpc.NewServer(opts...)

	resourceCache := cache.NewLinearCache(resource.SecretType, cache.WithLogger(wrapEnvoyLogger(s.logger.Named("cache"))))
	secretManager := NewSecretManager(s.client, resourceCache, s.logger.Named("secret-manager"))
	handler := NewRequestHandler(s.logger.Named("handler"), s.store, secretManager)
	sdsServer := server.NewServer(childCtx, resourceCache, handler)
	secretservice.RegisterSecretDiscoveryServiceServer(s.server, sdsServer)
	listener, err := net.Listen(s.protocol, s.bindAddress)
	if err != nil {
		return err
	}

	go secretManager.Manage(childCtx)
	errs := make(chan error, 1)
	go func() {
		errs <- s.server.Serve(listener)
	}()

	s.logger.Trace("running SDS server")
	for {
		select {
		case err := <-errs:
			return err
		case <-childCtx.Done():
			s.Shutdown()
			return nil
		case <-time.After(cachedMetricsTimeout):
			resources := len(resourceCache.GetResources())
			metrics.Registry.SetGauge(metrics.SDSCachedResources, float32(resources))
		}
	}
}

// Shutdown attempts to gracefully shutdown the server, it
// is called automatically when the context passed into the
// Run function is canceled.
func (s *SDSServer) Shutdown() {
	if s.server != nil {
		stopped := make(chan struct{})
		go func() {
			s.server.GracefulStop()
			close(stopped)
		}()

		timer := time.NewTimer(defaultShutdownTimeout)
		select {
		case <-timer.C:
			s.server.Stop()
		case <-stopped:
			timer.Stop()
		}
	}
}

func logFunc(log func(msg string, args ...interface{})) func(msg string, args ...interface{}) {
	return func(msg string, args ...interface{}) {
		log(fmt.Sprintf(msg, args...))
	}
}

func wrapEnvoyLogger(logger hclog.Logger) log.Logger {
	return log.LoggerFuncs{
		DebugFunc: logFunc(logger.Debug),
		InfoFunc:  logFunc(logger.Info),
		WarnFunc:  logFunc(logger.Warn),
		ErrorFunc: logFunc(logger.Error),
	}
}
