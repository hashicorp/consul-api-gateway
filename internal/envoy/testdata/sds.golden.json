{
  "admin": {
    "access_log_path": "/dev/null",
    "address": {
      "socket_address": {
        "address": "127.0.0.1",
        "port_value": 19000
      }
    }
  },
  "node": {
    "cluster": "",
    "id": "",
    "metadata": {
      "namespace": "default"
    }
  },
  "static_resources": {
    "clusters": [
      {
        "name": "self_admin",
        "ignore_health_on_host_removal": false,
        "connect_timeout": "5s",
        "type": "STATIC",
        "http_protocol_options": {},
        "loadAssignment": {
          "clusterName": "self_admin",
          "endpoints": [
            {
              "lbEndpoints": [
                {
                  "endpoint": {
                    "address": {
                      "socket_address": {
                        "address": "127.0.0.1",
                        "port_value": 19000
                      }
                    }
                  }
                }
              ]
            }
          ]
        }
      },
      {
        "name": "consul-server",
        "ignore_health_on_host_removal": false,
        "connect_timeout": "1s",
        "type": "STRICT_DNS",
        "http2_protocol_options": {},
        "loadAssignment": {
          "clusterName": "consul-server",
          "endpoints": [
            {
              "lbEndpoints": [
                {
                  "endpoint": {
                    "address": {
                      "socket_address": {
                        "address": "",
                        "port_value": 0
                      }
                    }
                  }
                }
              ]
            }
          ]
        }
      },
      {
        "kv": "here"
      }
    ],
    "listeners": [
      {
        "name": "envoy_ready_listener",
        "address": {
          "socket_address": {
            "address": "0.0.0.0",
            "port_value": 20000
          }
        },
        "filter_chains": [
          {
            "filters": [
              {
                "name": "envoy.filters.network.http_connection_manager",
                "typedConfig": {
                  "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
                  "stat_prefix": "envoy_ready",
                  "codec_type": "HTTP1",
                  "route_config": {
                    "name": "self_admin_route",
                    "virtual_hosts": [
                      {
                        "name": "self_admin",
                        "domains": [
                          "*"
                        ],
                        "routes": [
                          {
                            "match": {
                              "path": "/ready"
                            },
                            "route": {
                              "cluster": "self_admin",
                              "prefix_rewrite": "/ready"
                            }
                          },
                          {
                            "match": {
                              "prefix": "/"
                            },
                            "direct_response": {
                              "status": 404
                            }
                          }
                        ]
                      }
                    ]
                  },
                  "http_filters": [
                    {
                      "name": "envoy.filters.http.router"
                    }
                  ]
                }
              }
            ]
          }
        ]
      }
    ]
  },
  "dynamic_resources": {
    "lds_config": {
      "ads": {},
      "resource_api_version": "V3"
    },
    "cds_config": {
      "ads": {},
      "resource_api_version": "V3"
    },
    "ads_config": {
      "api_type": "DELTA_GRPC",
      "transport_api_version": "V3",
      "grpc_services": {
        "initial_metadata": [
          {
            "key": "x-consul-token",
            "value": ""
          }
        ],
        "envoy_grpc": {
          "cluster_name": "consul-server"
        }
      }
    }
  }
}
